### DISCUSSION TITLE
ベクトルタイルを南極に

### DISCUSSION THREAD

**Discussion by hfu** [optgeo/63ants#2] [Ideas]
とりあえず、国土地理院の「南極の地理空間情報」のうち Shapefile で提供されているものをベクトルタイルにしていくことについて情報を整理していきます。

**Comment by hfu** [comment]
## 目標とするデータの特定
1. 第一目標は [地図データ（SHAPE、KML、GML形式）](https://www.gsi.go.jp/antarctic/01.html)にある南極地形図だと思っています。
2. 第二目標として [デジタル標高データ（SHAPE形式）](https://www.gsi.go.jp/antarctic/13.html)も考慮可能と考えます。
3. その後、SCAR のデータについても検討する余地はあるかもしれません。

## 早速投影法について検討しておこう
[dirty-reprojectors](https://www.developmentseed.org/dirty-reprojectors-app/) の方法もあるとはいえ、ドームふじまでを含む日本の想定行動範囲であれば、まだウェブメルカトルで引っ張れると思っています。ズームレベルが「ズレる」漢字になるとは思っていますが、とりあえずは dirty-reprojector の方法を使わず、素のウェブ地図技術で対応することを試みます。

**Reply by SawagakiT** [reply]
こちらでもそのへんは試していて，昭和基地からDome-FまでのルートGPSトラックをUTM Zone37Sに投影したグリッドラスタ画像をQGISで作成済みです．この図のジオリファレンス済みのをkmzで書き出してgoogle Earthでよませると，いい感じに重なります．また，iOSの「スーパー地形」でも，OSMを下敷きにしてこのkmzをカスタムMapで重ねても，けっこういい感じに重なります．
![Dome_utm37s](https://user-images.githubusercontent.com/91312110/134689835-48d89f2a-9df5-406b-bd8a-6c5875cbc35d.jpg)

**Reply by SawagakiT** [reply]
昭和ーDomeF間はこれでも十分です．どうせ真っ白な雪原しかないので，歴代のドーム旅行隊が記録してきたGPSデータでナビゲーションできれば，とりあえずはOK．問題は基地設備や海岸線が入り組んでいる氷床縁辺部です．地理院のWebサイトでklmやgmlが公開されているのも，この氷床縁辺部エリアでして，それをベクタータイル化したい，というのが希望です．たぶんそのスケールならWebメルカトルでいけると思います．

**Reply by hfu** [reply]
ありがとうございます。まずは昭和基地、東西オングル島 2500 データですね。これを第一着手データとして、検討を進めてまいります。

**Comment by hfu** [comment]
# 早速ざっくりとベクトルタイル設計
## 2500
2500 のデータは、1996年のものと2015年以降のものがあるが、地域が重なっているようなので、2015年以降のものを使う。次の3つの ZIP ファイルからなる。

- 東オングル島周辺 https://www.gsi.go.jp/antarctic/contents/000153778.zip
- 西オングル島北東部 https://www.gsi.go.jp/antarctic/contents/000211385.zip
- 西オングル島南東部 https://www.gsi.go.jp/antarctic/contents/000211446.zip

![image](https://user-images.githubusercontent.com/18297/134692113-520b031c-f1f7-4550-b096-6c06dc65c335.png)

ざっくり、このデータを例えば z=13 あるいは 12 くらいに入れるのではないか。

## 25000
25000 は 1959～1983年 作成らしく古い。古いが、他にない内容を持っている。

- プリンスオラフ https://www.gsi.go.jp/antarctic/contents/000113450.zip
- 昭和基地周辺 https://www.gsi.go.jp/antarctic/contents/000113722.zip
- リュツォホルム https://www.gsi.go.jp/antarctic/contents/000113733.zip
- プリンスハラルド https://www.gsi.go.jp/antarctic/contents/000113614.zip
- やまと山脈 https://www.gsi.go.jp/antarctic/contents/000113538.zip

![image](https://user-images.githubusercontent.com/18297/134692473-f654c7e8-12d8-42bf-87af-a99c93e8ce2f.png)

## 50000
当初作成と、楕円体高版と、標高版があるが、整備地域が重なっているようなので、最新の標高版（作成2014年〜）を採用する。

### セールロンダーネ
- セールロンダーネ https://www.gsi.go.jp/antarctic/contents/000115280.zip

### ベルジカ山地・やまと山脈
- やまと山脈 https://www.gsi.go.jp/antarctic/contents/000211395.zip
- ベルジカ山地 https://www.gsi.go.jp/antarctic/contents/000211394.zip

### リーセル・ラルセン山
- リーセル・ラルセン山 https://www.gsi.go.jp/antarctic/contents/000211447.zip

**Reply by SawagakiT** [reply]
昭和基地の建物や観測アンテナの位置まで確認できるようにするなら2015年以降版のZ=17くらいが必要で，たぶんそこまではいけます．

**Reply by hfu** [reply]
まだ空ですが、試作用のレポジトリを作りました。
https://github.com/optgeo/a-1

**Reply by SawagakiT** [reply]
南極観測隊が現地で実際に使うシーンのイメージをつかんでいただくために，これまで私の方で構築したWebアプリを下記に公開します．私たちが南極に出発する前の10月末まで閲覧できます．
https://tv.jare.0g0.jp/c2.html

これは地理院作成の東オングル島周辺2500のkmlデータからいくつかの主要要素を抜き出して，それをleafletというjsでWebアプリ化してラズパイのapacheから送出しています．kmlをそのまま描画していおりタイル化はしていません．さらにこれに，現在昭和基地で越冬中の62次観測隊が開拓したばかりの雪上車走行ルートのGPSトラックを重ね，大陸氷床の表面状態がわかるように人工衛星画像もkmzラスタタイルにして重ねています．なお，ベースの地図はOSMです．

一応，スマホのブラウザで見たときには，スマホ内蔵のGPSと連携して現在地も表示できるようにはしてありますが，なにせまだ現地にはいないので，はたしてどこまで正確にポジショニングできているかは未確認です．確認は実際に昭和基地に行ってからのお楽しみです．

このように，南極観測隊の活動現場では，メートルスケールの基地管理情報から，数kmスケールのルート情報，そして数百kmスケールの氷床表面データまでを同時に扱います．衛星画像はラスタのまま使わざるを得ませんが，基地の建物・道路・雪上車のルートなどの地理情報をベクトルタイルにして，インターネットが使えない場所でもオンプレミスで使えるようにするのが最終目標となります．そのシステムとしてUNVTに期待しているところです．

**Comment by hfu** [comment]
とりあえず、2500 等高線を変換してみました！

https://x.optgeo.org/a-1/#12.08/-69.01874/39.54273

![ 2021-09-26 23 13 57](https://user-images.githubusercontent.com/18297/134811543-c8c186d7-9525-4393-833a-242339f54cbc.jpg)

**Reply by hfu** [reply]
南極ベクトルタイルの難しさは、ズームレベルが中緯度帯と比べてズレることかもしれません。ズームレベル12でかなり大きな縮尺になっている気がします。

**Reply by SawagakiT** [reply]
さっそくにどうも．2mコンターですね．高緯度は実距離に対して経度がつまってきますからズームレベルの話もなるほどです．私的にはこっちのほうが普通なので「基地設備管理レベルならZ=16〜17かな」という感覚でした．ちなみに，海氷・積雪域と露岩域とが地形的に連続するため0m汀線が明瞭にならないのもこの地域の特徴です．じつは基準高も2015前後でジオイドを考慮するかどうかで変更されてますので，そもそも０ｍ自体が未確定なのですけど，それもまたこの図幅の特徴でもあります．

**Comment by ubukawa** [comment]
お世話になります。国連地理空間情報課の鵜生川です。
国連事務局でベクトルタイルの実装に取組んでいます。

我々のところではまだラズベリーパイを十分に活用できてないですが、ラズベリーパイでのベクトルタイルのホスティングに興味があります。
南極でのタイル配信のためのホスティングはbudoを考えられているのでしょうか？

（ローカルなウェブサーバーを立ち上げるにしてもタイル形式がpbfなので、onyxのようなサーバがなくてもnginxのようなもので十分なのかなぁと想像しています。）

※自分の練習用に、onyxに静的なホスティングを足して、マップライブラリとspriteなどを加えたレポジトリを以下に作ってみました。
https://github.com/ubukawa/b-quartz

**Reply by SawagakiT** [reply]
鵜生川さま　はじめまして
南極でのベクタータイル配信を現場で実現したくて，いろいろ妄想しております澤柿です．
私はまだ勉強途中で全体像が見えていません．まず最初にとびついたのがmisoraで，これでできあいのタイルマップを配信することろまではなんとかできるようになりました．UNVTにはonyxという別のベクタルタイルサーバーもあるようですが，githubにある情報だけでは素人にはまだ難しいです．

なおmiroraとは別になりますが，下記のサイトのやりたかをそのままなぞって，OSMの公開データセットから自前でベクトルタイルを作成してホストするところまではラズパイで構築済みです．これはapacheにmod_tileというモジュールを組み込んでタイルリクエストを処理しているようですね（この理解であってますか？）．
https://switch2osm.org/serving-tiles/manually-building-a-tile-server-18-04-lts/

また，https://tv.jare.0g0.jp/c2.html　で昭和基地周辺のWebGISの習作を公開していますが，このバックに使っているOSM Mapも，上記の方法で作成しホスティングしているものになります．

それで私がまだちゃんと理解できていないのは，ベクトルタイルのデータ生成方法のほうです．pbfタイル形式までもっていけさえすれば，あとはapache＋mod_tileであれnginxであれ，配信はどうにでもなるように理解しているのですが，ここも素人理解なので，正確なところが知りたいと思っています．

**Reply by hfu** [reply]
@ubukawa タイルのホスト方法は、UNVTの他の要素である生産・スタイル・最適化・利用と同じく自由ですが、私は次のように考えています。

- 定常運用する場合には、なるべく一般的に使われているものを、できるだけ手を入れることなく、サービスとして使うのが良いと思っています。私自身は `nginx` を使っています。
- 他方で、開発作業の中での動作確認に `nginx` を使おうとすると、`nginx` 環境が「汚れていく」傾向があるので、開発作業の中では `budo` を使っています。

@SawagakiT 先生のおっしゃるとおり、「配信はどうにでもなる」のがタイル技術の美点でもあり、配備環境の都合に合わせて動けるフレキシビリティを確保するのが UNVT の基本的な戦略になると思っています。

その上で、何らかの「簡単パッケージ」を組めることは第二目標としては良いと思っていますが、だいたいにおいて「簡単パッケージ」は本格的には使えないということになることが多いです。国土地理院や UNGSC における本格的な運用をコアのユースケースのひとつとして捉えている以上は、第一に本格環境で間違いなく使えること、第二に様々な環境で使える柔軟性を持つこと、という順番で物事を考えることになると思っています。

# 現在進めている作業
私の方では、現在のところ次のような作業を進めています。

1. ベクトルタイルのスタイルを効率的に記述するツールである https://github.com/geolonia/charites を https://github.com/optgeo/a-1 で活用していくつもりです。
2. そのために、まずは https://github.com/unvt/equinox に charites を組み込む作業を進めています。
3. これと同時に、https://github.com/optgeo/63ants/discussions/3 の文脈の中で、Raspberry Pi で GPS アンテナを使うためのソフトウェアも equinox に含めておこうと思っています。

**Reply by ubukawa** [reply]
@SawagakiT  さま、早速コメントありがとうございます。
私も最初は全体像がなかなか見えなかったのですが、藤村様に教えていただきつつ国連で何とか作業を進めているところです。どうぞよろしくお願いいたします。
onyxは私も最初はさっぱりわからなかったので、難しいのもよくわかります。（私は以下の点を理解するのに時間がかかりました）
　・サーバーをnodejsで作っている。
　・これはおそらくmbtilesからpbfを効率よく配信するためにnodejsを使っている。（app.jsにはこのためのことがたくさん書いてありますが、小縮尺のタイル区画ごとにmbtilesを作っていることを知らないと、ここでやっていることが理解できにくいと思います。→例えば、6-37-31.mbtilesというファイルがあって、これに7-74-62.pbfなどが入っている。）
　・pm2でnodejsをデーモンプロセスにしてサーバーを走らせている。
　・httpsにするための認証をとってこないといけない。

また、mod_tileの方法も教えていただきありがとうございます。
私はapacheでmod_tileを使ったことはありません。
mod_tileはmaputnikをつかってレンダリングしていたり、データはpostgisとして持っていたりするようで、いろいろな方法があるのだなぁと思いました。（データが大きいとサーバーの負荷が大きそうなので私の用途には向かなさそうです。）

ベクトルタイルの作成については、pbfになってしまえば配信はどうにもなると思います。

ベクタータイルの作成はTippecanoeというツール（インプットがgeojson形式でアウトプットがベクトルタイル）を使うのが一番便利だと思います。UNVTのequinoxでもTippecanoeを生産に使っています。
ただし、tippecanoeはLinuxのような環境でしか動かないので、Macだとターミナル、WindowsだとWSLやDocker、あるいはRaspberry Piで使うということなので、一般的なGISツールに比べてハードルが高いかもしれません。それに加えて以下にソースデータを入力形式であるGeoJSONにするかという工夫が各ツールでやり方が少し違うと思います。
昔、私がWindowsのDocker環境でベクトルタイルを作る練習をしたスライドがありますので、流れを見るのに必要でしたらご参照ください。
https://speakerdeck.com/ubukawa/2021-05-28-vector-tile-conversion-with-doker-unvt-nanban
（南極の実装は、現在藤村様がやっているように、ラズベリーパイで生産とホスティングとスタイルを一緒にやってしまった方が効率的だと思います。）

@hfu さま、nginxとのことをおしえていただきありがとうございました。budoを今使っているわけもよくわかりました。
私のところでも将来ラズベリーパイを使う機会があれば、mbtilesでなくてpbfに展開して保存しておくことで一般的に使われているnginxでいけるなぁと想像しました。（もちろんそのようなケースがあるかは現場のニーズ調査が必須ですが、その前に自分の中でいろいろ想像するのに役立ちます。）

よこからお邪魔をしましたが、いろいろを教えていただきありがとうございました。

**Comment by hfu** [comment]
少し間が空きましたが、ベクトルタイルのスタイルをアップデートしました。

https://github.com/unvt/charites という新しいツールを使って、よりエレガントにスタイルをつけることができるようになってきています。

# デモサイト
https://x.optgeo.org/a-1/#12.67/-69.01099/39.57001

![ 2021-10-17 17 30 38](https://user-images.githubusercontent.com/18297/137618827-6d2d1825-1cb5-43c1-94a6-6d488ec992bf.jpg)

# コード及びドキュメント
https://github.com/optgeo/a-1

**Comment by hfu** [comment]
a-1 を作り進めて、ここまで作りました！https://x.optgeo.org/a-1/#16.05/-69.004583/39.582096

![ 2021-11-21 10 14 05](https://user-images.githubusercontent.com/18297/142745534-7fb83d28-173f-4521-bb57-5c2f512f1ad2.jpg)
